### PR TITLE
Convert DetektJvmSpec to use ProjectBuilder

### PR DESCRIPTION
**Pull request by 3flex**
See #3778. This reduced execution of the DetektJvmSpec test by about 25% on my machine.

### CODE REVIEW

**Comment by 3flex** [comment]
> Is it fair to assume that the improvement comes from changing from groovy interpreter to Kotlin compiler?

It's from using [`ProjectBuilder`](https://docs.gradle.org/current/javadoc/org/gradle/testfixtures/ProjectBuilder.html) instead of Gradle's TestKit.

TestKit runs a full Gradle build and spins up its own daemon. ProjectBuilder is much more limited, and can't be used for functional tests, because you can't execute tasks. But for checking things like whether task input & output values are set correctly it's suitable, and faster.

**Review by chao2zhang — APPROVED** [review]

**Review by schalkms — APPROVED** [review]

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt [hunk 1]
```diff
@@ -13,9 +15,17 @@ class DslGradleRunner @Suppress("LongParameterList") constructor(
...
     val gradleVersionOrNone: String? = null,
     val dryRun: Boolean = false
 ) {
+    constructor(
+        projectLayout: ProjectLayout,
+        buildFileName: String,
+        projectScript: Project.() -> Unit,
+        configFileOrNone: String? = null,
+        baselineFiles: List<String> = emptyList(),
+    ) : this(projectLayout, buildFileName, "", configFileOrNone, baselineFiles, projectScript)
```
Comment (cortinico) [inline detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:28]: This is confusing to me. Why do we need an extra ctor where the new paramters is in 3rd position?

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt [hunk 2]
```diff
@@ -13,9 +15,17 @@ class DslGradleRunner @Suppress("LongParameterList") constructor(
     val mainBuildFileContent: String,
     val configFileOrNone: String? = null,
     val baselineFiles: List<String> = emptyList(),
+    val projectScript: (Project.() -> Unit)? = null,
```
Comment (cortinico) [inline detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:18]: Why this is not last? You could benefit also of trailing lambda syntax

Comment (cortinico) [reply detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:18]: Also does it need to be nullable? Can it just be an empty lambda?

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt [hunk 3]
```diff
@@ -109,6 +119,18 @@ class DslGradleRunner @Suppress("LongParameterList") constructor(
     private val Submodule.moduleRoot: File
         get() = File(rootDir, name).apply { mkdirs() }
 
+    fun buildProject(): Project {
+        val project = ProjectBuilder.builder()
+            .withProjectDir(rootDir)
+            .build()
+
+        requireNotNull(projectScript)
```
Comment (cortinico) [inline detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:127]: You can remove this if `projectScript` becomes non nullable.

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt [hunk 4]
```diff
@@ -109,6 +119,18 @@ class DslGradleRunner @Suppress("LongParameterList") constructor(
     private val Submodule.moduleRoot: File
         get() = File(rootDir, name).apply { mkdirs() }
 
+    fun buildProject(): Project {
```
Comment (cortinico) [inline detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:122]:
```suggestion
    fun buildProject(): Project = ProjectBuilder.builder()
        .withProjectDir(rootDir)
        .build()
        .apply { projectScript.invoke(this) }
```

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektJvmSpec.kt [hunk 5]
```diff
@@ -13,45 +19,48 @@ object DetektJvmSpec : Spek({
...
-            gradleRunner.runTasksAndCheckResult(":detektMain") { buildResult ->
-                assertThat(buildResult.output).containsPattern("""--baseline \S*[/\\]detekt-baseline-main.xml """)
-                assertThat(buildResult.output).contains("--report xml:")
-                assertThat(buildResult.output).contains("--report sarif:")
-                assertThat(buildResult.output).doesNotContain("--report txt:")
-                assertThat(buildResult.output).contains("--classpath")
-            }
+            val project = gradleRunner.buildProject()
+
+            project.getTasksByName("detektMain", true) // trigger project evaluation
```
Comment (cortinico) [inline detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektJvmSpec.kt:39]: Is the test not working without this? Isn't the `getByPath` below sufficient?

**Review by cortinico — APPROVED** [review]
Great work 👍 

Just one side note: now we're moving from actually running Gradle tasks to just setting up a project and asserting that the tasks are configured properly.

Not a big deal but if the task blows up during execution, the former task would have found it. This one not. Something to keep in mind 👍

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/testkit/DslGradleRunner.kt:18]: Good catch

File: detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektJvmSpec.kt (hunk 5 above)
Comment (3flex) [reply detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektJvmSpec.kt:39]: Unfortunately not - currently `project.afterEvaluate` is used throughout the detekt plugin to register tasks, so without triggering the project evaluation in the test, getByPath throws UnknownTaskException.

Comment (cortinico) [reply detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/DetektJvmSpec.kt:39]: How do you feel about having a:

```
fun Project.triggerEvaluation() { project.getTasksByName("tasks", true) }
```

**Review by BraisGabin — APPROVED** [review]

**Review by chao2zhang — APPROVED** [review]

**Pull request merged**